### PR TITLE
use exact matching of allowed domain entries, issue #489

### PR DESCRIPTION
**Pull request by emicklei**

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
The docs should be updated to indicate that each allowed domain isn't a _potential_ regexp pattern, it _is_ a regexp pattern. So `^some.example.com$` matches `somexexample.com`. Even after the PR, `^some\.example\.com` matches `some.example.com.org`.

**Comment by emicklei** [comment]
@JamieSlome is your hunt issue resolved by this PR?

File: cors_filter.go [hunk 1]
```diff
@@ -21,7 +21,11 @@ import (
 type CrossOriginResourceSharing struct {
 	ExposeHeaders  []string // list of Header names
 	AllowedHeaders []string // list of Header names
-	AllowedDomains []string // list of allowed values for Http Origin. An allowed value can be a regular expression to support subdomain matching. If empty all are allowed.
+	// AllowedDomains list of allowed values for Http Origin.
+	// An allowed value can be a regular expression to support subdomain matching.
+	// Non-regular expression values will be changed into an exact match: ^yourdomain.com$
```
Comment (crenshaw-dev) [inline cors_filter.go:26]: I'm not sure `^yourdomain.com$` is an intuitive "exact match". The `.` is still a wildcard. You'd have to do `fmt.Sprintf("^%s$", regexp.QuoteMeta(each))` to get an "exact match".

Comment (crenshaw-dev) [inline cors_filter.go:26]: It's not clear to me from this what qualifies as "non-regular expression". `example.com` and `.*` and `example\.com` are all regular expressions, even though they don't start with `^` or end with `$`.

**Comment by JamieSlome** [comment]
@emicklei:

Just (cc) @sickcodes and @ktg9 to see if they think this addresses the reported issue 👍

File: cors_filter.go (hunk 1 above)
Comment (steven-collins-omega) [reply cors_filter.go:26]: agreed. `^www.amazon.com$` still matches `wwwqamazon.com` and `www.amazonqcom`.

**Comment by steven-collins-omega** [comment]
@emicklei if I'm reading the code correctly, this just simply isn't going to work in practice in most cases. `isOriginAllowed()` is called on the literal value of the Origin header, [which looks something like](https://html.spec.whatwg.org/multipage/origin.html#ascii-serialisation-of-an-origin) `https://some.example.com:8080`. Technically the documentation does say that `AllowedDomains` is a "list of allowed values for Http Origin", which is what it is. However, given the name `AllowedDomains` and the `cors_filter` test code, a library user would be likely to interpret it as a list of allowed domains. Such a user will find, after this proposed change, that nothing is allowed anymore. `^some.example.com$` doesn't match `https://some.example.com:8080`.

**Comment by emicklei** [comment]
@steven-collins-omega thank you for commenting on this. I agree that it is not correct and confusing. I will give it second time to come up with a solution

**Comment by emicklei** [comment]
hi all, I thinking about changing the behaviour of CrossOriginResourceSharing:

- AllowedDomains will have entries that must include the origin value from the request by string comparing.
- add the AllowedDomainFunc field which can be set with a function takes takes the `origin` and returns a bool for a match.  This allows the developer define its own logic.

File: cors_filter.go [hunk 2]
```diff
@@ -120,35 +133,22 @@ func (c CrossOriginResourceSharing) isOriginAllowed(origin string) bool {
...
 		return true
 	}
 
-	allowed := false
+	// exact match on each allowed domain
 	for _, domain := range c.AllowedDomains {
-		if domain == origin {
-			allowed = true
-			break
+		if domain == ".*" || domain == origin {
```
Comment (crenshaw-dev) [inline cors_filter.go:144]: Should `domain` and `origin` be lowercased before comparison?

Comment (emicklei) [reply cors_filter.go:144]: yes, i think this should happen.  RFC4343: Domain Name System (DNS) names are "case insensitive".

**Pull request merged**